keps/sig-network/20191104-iptables-no-cluster-cidr.md

```diff
@@ -0,0 +1,358 @@
+
+   1. [Masquerade off cluster traffic to services by node IP](https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/proxier.go#L1558-L1563)
+   2. [Accepting traffic after first packet, after being accepted by kubernetes](https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/proxier.go#L1635-L1654)
+
+This enhancement proposes ways to achieve similar goals without tracking the pod cluster CIDR to do so.
+
+## Motivation
+
+The idea that makes kubernetes networking model unique and powerful is the concept of each pod having its own IP, 
+with all the pod IPs being natively routable within the cluster. The service chains in iptable rules depend on this 
+capability by assuming that they can treat all the endpoints of a cluster as being equivalent and load balance service 
+traffic across all the endpoints, by just translating destination to the pod IP address.
+
+While this is powerful, it also means pod IP addresses are in many cases the constraining resource for cluster creation
+and scale. It would be valuable for implementations to have different strategies for managing pod IP addresses that can
+adapt to different environment needs.
+
+Some examples of use cases:
+
+   * Creating a cluster out of many disjoint ranges instead of a single range.
+   * Expanding a cluster with more disjoint ranges after initial creation.
+
+Not having to depend on the cluster pod CIDR for routing service traffic would effectively de-couple pod IP management
+and allocation strategies from service management and routing. Which in turn would mean that it would be far cheaper 
+to evolve the IP allocation schemes while sharing the same service implementation, thus significantly lowering the bar
+for adoption of alternate schemes.
+
+Alternate implementations that don’t use iptables could also adopt this same reasoning to not have to track the cluster
+CIDR for routing cluster traffic.
+
+### Goals
+
+   * Not having to depend on the cluster pod CIDR for iptable rules and cluster traffic routing.
+
+### Non-Goals
+
+   * Providing alternate models of IP allocation schemes for pod CIDR.
+   * Enhancing current allocators to handle disjoint ranges.
+   * Enhancing current allocators to add additional ranges after cluster creation.
+   * Changing current assumptions around having a single pod CIDR per node.
+
+## Proposal
+
+As stated above, the goal is to re-implement the functionality called out in the summary, but in a 
+way that does not depend on a pod cluster CIDR. The essence of the proposal is that for the 
+first two cases in iptables implementation and first case in ipvs, we can replace the `-s proxier.clusterCIDR` with 
+some notion of node local pod traffic.
+
+The core logic in these cases is “how to determine” cluster originated traffic from non-cluster originated ones. 
+The proposal is that tracking pod traffic generated from within the node is sufficient to determine cluster originated 
```

Comment:
I think you're making two assumptions here that don't hold for all clusters. The first is that all non-cluster-originated traffic cannot reach the cluster - this is often false, especially for Calico (cc @caseydavenport). The second one is that node-originated traffic is the complete set of traffic that does not need masquerading.

Reply:
I don't think this is about reachability. The more I think about it, I think it's about making sure that the return path from the backend pod hits the node where the DNAT happened so that the reverse DNAT can happen on the way back. Without this the end-to-end connection breaks. Using cluster pod-cidr to do this work today because we do the DNAT on every node boundary, and so the return path will hit that node. In case we don't do the DNAT on every node boundary, then we have to SNAT all traffic that did not originate from the node (even if part of the cluster pod-cidr) as otherwise the return path will not hit that node. I think that is what we are trying to reason through. And from that perspective, it seems to make more and more sense to write the rule as "locally generated from the node - no masquerad" vs "not locally generated from the node - so masquerade". Right?

Reply:
@squeed based on this understanding, can we mark this as resolved?

Reply:
@squeed I think your first point is wrong (that this assumes ll non-cluster-originated traffic cannot reach the cluster).  That is certainly possible in GCP, for example, and I don't see how this proposal changes that.

I think your second point is correct (that this assumes that node-originated traffic is the complete set of traffic that does not need masquerading).  Can you expand on that?

Reply:
@thockin (sorry, k/enhancements mentions got blackholed). Argh, I forgot about NAT + return path. Ignore the noise, sorry.